src/main/java/io/airlift/slice/Slice.java

```diff
@@ -815,7 +838,13 @@ public Slice slice(int index, int length)
         if (length == 0) {
             return Slices.EMPTY_SLICE;
         }
-        return new Slice(base, address + index, length, retainedSize, reference);
+
+        if (reference == COMPACT) {
+            return new Slice(base, address + index, length, retainedSize, NOT_COMPACT);
+        }
+        else {
```

Comment:
Nit: the `else` is redundant, remove it (thus the original line will remain unchanged)